File: src/main/java/net/slipp/support/utils/RankingUtils.java
package net.slipp.support.utils;

import java.util.Calendar;

public class RankingUtils {
    public static double calculateHotScore(int comments, int likes, Calendar standard, Calendar date) {
        int score = score(comments, likes);
        double order = Math.log10(Math.max(Math.abs(score), 1));
        long seconds = epochSeconds(standard, date) - 1134028003;
        double result = order + seconds / 45000;
        return Math.round(result * 10000000.0) / 10000000.0;
    }

    private static int score(int comments, int likes) {
        return comments + likes;
    }

    static long epochSeconds(Calendar standard, Calendar date) {
        long differenceMillis = date.getTimeInMillis() - standard.getTimeInMillis();
        return differenceMillis / 1000;
    }

    public static double calculateCommentRanking(int ups, int downs) {
        int number = ups + downs;
        if (number == 0) {
            return 0;
        }

        int n = ups + downs;
        double z = 1.0;
        float phat = new Integer(ups).floatValue() / n;
        return Math.sqrt(phat + z * z / (2 * n) - z * ((phat * (1 - phat) + z * z / (4 * n)) / n)) / (1 + z * z / n);
    }
}

File: src/main/scala/net/slipp/support/utils/RankingUtils.scala
package net.slipp.support.utils

import java.util.Calendar

object RankingUtils {
  def calculateHotScore(comments: Int, likes: Int, standard: Calendar, date: Calendar): Double = {
    val score = calculateScore(comments, likes)
    val order: Double = Math.log10(Math.max(Math.abs(score), 1))
    val seconds: Long = epochSeconds(standard, date) - 1134028003
    val result: Double = order + seconds / 45000
    return result * 10000000.0.round / 10000000.0
  }

  private def calculateScore(comments: Int, likes: Int): Int = {
    return comments + likes
  }

  private[utils] def epochSeconds(standard: Calendar, date: Calendar): Long = {
    val differenceMillis: Long = date.getTimeInMillis - standard.getTimeInMillis
    return differenceMillis / 1000
  }

  def calculateCommentRanking(ups: Int, downs: Int): Double = {
    val number: Int = ups + downs
    if (number == 0) {
      return 0
    }
    val n: Int = ups + downs
    val z: Double = 1.0
    val phat: Float = new Integer(ups).floatValue / n
    return Math.sqrt(phat + z * z / (2 * n) - z * ((phat * (1 - phat) + z * z / (4 * n)) / n)) / (1 + z * z / n)
  }
}

